### ISSUE TITLE
copyfd_io_poll: poll timed out: race/bug with poll wakeup

### ISSUE THREAD

**Issue by matttbe** [multipath-tcp/mptcp_net-next#147]
More regularly now, my CI is reporting issues with  `copyfd_io_poll: poll timed out`.

It is generally when starting the first selftests:

```
01:00:25.182 # selftests: net/mptcp: mptcp_connect.sh
01:00:25.330 [   62.034357] IPv6: ADDRCONF(NETDEV_CHANGE): ns1eth2: link becomes ready
01:00:26.995 [   62.483523] IPv6: ADDRCONF(NETDEV_CHANGE): ns2eth1: link becomes ready
01:00:27.444 [   62.697470] IPv6: ADDRCONF(NETDEV_CHANGE): ns3eth2: link becomes ready
01:00:27.658 [   62.699200] IPv6: ADDRCONF(NETDEV_CHANGE): ns2eth3: link becomes ready
01:00:27.660 [   63.357947] IPv6: ADDRCONF(NETDEV_CHANGE): ns4eth3: link becomes ready
01:00:28.319 [   63.359382] IPv6: ADDRCONF(NETDEV_CHANGE): ns3eth4: link becomes ready
01:00:28.320 # INFO: set ns3-600fb815-1a9Kow dev ns3eth2: ethtool -K tso off gro off
01:00:28.625 # Created /tmp/tmp.xXIaPgnXs8 (size 6890524	/tmp/tmp.xXIaPgnXs8) containing data sent by client
01:00:29.080 # Created /tmp/tmp.N8TwHJfNyq (size 5486620	/tmp/tmp.N8TwHJfNyq) containing data sent by server
01:00:29.396 # New MPTCP socket can be blocked via sysctl		[ OK ]
01:00:29.816 # setsockopt(..., TCP_ULP, "mptcp", ...) blocked	[ OK ]
01:00:30.004 # INFO: validating network environment with pings
01:00:30.060 [   70.135700] netem: version 1.3
01:00:35.096 # INFO: Using loss of 0.52% delay 23 ms reorder 93% 22% with delay 5ms on ns3eth4
01:00:35.113 # ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP	(duration   301ms) [ OK ]
01:00:36.589 # ns1 MPTCP -> ns1 (10.0.1.1:10001      ) TCP  	(duration   217ms) [ OK ]
01:00:37.940 # ns1 TCP   -> ns1 (10.0.1.1:10002      ) MPTCP	(duration   215ms) [ OK ]
01:00:39.283 # ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
01:01:10.231 # (duration 30368ms) [ FAIL ] client exit code 2, server 0
01:01:10.290 # 
01:01:10.290 # netns ns1-600fb815-1a9Kow socket stat for 10003:
01:01:10.291 # State Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
01:01:10.787 # ESTAB 0      0      [dead:beef:1::1]:10003 [dead:beef:1::1]:37334
01:01:10.788 # 	 ts sack cubic wscale:7,7 rto:201 rtt:0.289/0.088 ato:40 mss:32768 pmtu:65536 rcvmss:65436 advmss:65464 cwnd:10 bytes_sent:4526163 bytes_acked:4526163 bytes_received:6890524 segs_out:408 segs_in:340 data_segs_out:320 data_segs_in:280 send 9070726644bps lastsnd:30573 lastrcv:30573 lastack:30573 pacing_rate 18110120896bps delivery_rate 1180828824bps delivered:321 busy:174ms rwnd_limited:3ms(1.7%) sndbuf_limited:13ms(7.5%) rcv_rtt:0.884 rcv_space:65464 rcv_ssthresh:1898120 minrtt:0.089 tcp-ulp-mptcp flags:Mec token:0000(id:0)/262a3b9e(id:0) seq:35a86fb92842ba47 sfseq:692001 ssnoff:8107faed maplen:41c
01:01:10.796 # 
01:01:10.801 # netns ns1-600fb815-1a9Kow socket stat for 10003:
01:01:10.802 # State Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
01:01:10.896 # ESTAB 0      0      [dead:beef:1::1]:37334 [dead:beef:1::1]:10003
01:01:10.897 # 	 ts sack cubic wscale:7,7 rto:201 rtt:0.705/0.125 ato:40 mss:65464 pmtu:65536 rcvmss:32768 advmss:65464 cwnd:22 ssthresh:22 bytes_sent:6890524 bytes_acked:6890525 bytes_received:4526163 segs_out:340 segs_in:409 data_segs_out:280 data_segs_in:320 send 16342785816bps lastsnd:30695 lastrcv:30695 lastack:30695 pacing_rate 19607866400bps delivery_rate 787536840bps delivered:281 busy:167ms rcv_rtt:0.459 rcv_space:65476 rcv_ssthresh:65476 minrtt:0.096 tcp-ulp-mptcp flags:Mmec token:0000(id:0)/7eac7c5c(id:0) seq:d4b7487fda3b268d sfseq:450fd4 ssnoff:79893660 maplen:80
01:01:10.904 # FAIL: Could not even run loopback v6 test
01:01:10.929 not ok 1 selftests: net/mptcp: mptcp_connect.sh # exit=1
```

But not always.

I get them more regularly now but not all the time. Here are some logs from August, grepping for `copyfd_io_poll` and `Could not even run loopback`:

```
export/20200827T050630
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 4)
# FAIL: Could not even run loopback test

export/20200904T170444
# ns2 MPTCP -> ns4 (10.0.3.1:10022      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20200910T202702
# ns3 MPTCP -> ns2 (dead:beef:2::1:10029) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20200917T170602
# ns2 MPTCP -> ns3 (dead:beef:2::2:10019) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20200923T081152
# ns4 MPTCP -> ns2 (10.0.1.2:10034      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)

export/20201002T160554
# ns3 MPTCP -> ns2 (10.0.1.2:10026      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20201204T192621
# ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20210113T180529
# ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 0, POLLOUT 4)
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback v6 test

export/20210116T091930
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# copyfd_io_poll: poll timed out (events: POLLIN 0, POLLOUT 4)

export/20210116T100943
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# copyfd_io_poll: poll timed out (events: POLLIN 0, POLLOUT 4)
# ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20210119T062117
# ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test
# FAIL: Could not even run loopback test
# FAIL: Could not even run loopback test
# FAIL: Could not even run loopback test

export/20210120T143433
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# copyfd_io_poll: poll timed out (events: POLLIN 0, POLLOUT 4)

export/20210121T064902
# ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback v6 test

export/20210123T081839
# ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback test

export/20210126T061955
# ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# FAIL: Could not even run loopback v6 test

export/20210127T062158
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# copyfd_io_poll: poll timed out (events: POLLIN 0, POLLOUT 4)

export/20210131T062218
# copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# copyfd_io_poll: poll timed out (events: POLLIN 0, POLLOUT 4)
```

**Comment by pabeni** [comment]
> More regularly now, my CI is reporting issues with `copyfd_io_poll: poll timed out`.
> It is generally when starting the first selftests:
> ```
> 01:00:25.182 # selftests: net/mptcp: mptcp_connect.sh
> 01:00:25.330 [   62.034357] IPv6: ADDRCONF(NETDEV_CHANGE): ns1eth2: link becomes ready
> 01:00:26.995 [   62.483523] IPv6: ADDRCONF(NETDEV_CHANGE): ns2eth1: link becomes ready
> 01:00:27.444 [   62.697470] IPv6: ADDRCONF(NETDEV_CHANGE): ns3eth2: link becomes ready
> 01:00:27.658 [   62.699200] IPv6: ADDRCONF(NETDEV_CHANGE): ns2eth3: link becomes ready
> 01:00:27.660 [   63.357947] IPv6: ADDRCONF(NETDEV_CHANGE): ns4eth3: link becomes ready
> 01:00:28.319 [   63.359382] IPv6: ADDRCONF(NETDEV_CHANGE): ns3eth4: link becomes ready
> 01:00:28.320 # INFO: set ns3-600fb815-1a9Kow dev ns3eth2: ethtool -K tso off gro off
> 01:00:28.625 # Created /tmp/tmp.xXIaPgnXs8 (size 6890524	/tmp/tmp.xXIaPgnXs8) containing data sent by client
> 01:00:29.080 # Created /tmp/tmp.N8TwHJfNyq (size 5486620	/tmp/tmp.N8TwHJfNyq) containing data sent by server
> 01:00:29.396 # New MPTCP socket can be blocked via sysctl		[ OK ]
> 01:00:29.816 # setsockopt(..., TCP_ULP, "mptcp", ...) blocked	[ OK ]
> 01:00:30.004 # INFO: validating network environment with pings
> 01:00:30.060 [   70.135700] netem: version 1.3
> 01:00:35.096 # INFO: Using loss of 0.52% delay 23 ms reorder 93% 22% with delay 5ms on ns3eth4
> 01:00:35.113 # ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP	(duration   301ms) [ OK ]
> 01:00:36.589 # ns1 MPTCP -> ns1 (10.0.1.1:10001      ) TCP  	(duration   217ms) [ OK ]
> 01:00:37.940 # ns1 TCP   -> ns1 (10.0.1.1:10002      ) MPTCP	(duration   215ms) [ OK ]
> 01:00:39.283 # ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP	copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
> 01:01:10.231 # (duration 30368ms) [ FAIL ] client exit code 2, server 0
> 01:01:10.290 # 
> 01:01:10.290 # netns ns1-600fb815-1a9Kow socket stat for 10003:
> 01:01:10.291 # State Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
> 01:01:10.787 # ESTAB 0      0      [dead:beef:1::1]:10003 [dead:beef:1::1]:37334
> 01:01:10.788 # 	 ts sack cubic wscale:7,7 rto:201 rtt:0.289/0.088 ato:40 mss:32768 pmtu:65536 rcvmss:65436 advmss:65464 cwnd:10 bytes_sent:4526163 bytes_acked:4526163 bytes_received:6890524 segs_out:408 segs_in:340 data_segs_out:320 data_segs_in:280 send 9070726644bps lastsnd:30573 lastrcv:30573 lastack:30573 pacing_rate 18110120896bps delivery_rate 1180828824bps delivered:321 busy:174ms rwnd_limited:3ms(1.7%) sndbuf_limited:13ms(7.5%) rcv_rtt:0.884 rcv_space:65464 rcv_ssthresh:1898120 minrtt:0.089 tcp-ulp-mptcp flags:Mec token:0000(id:0)/262a3b9e(id:0) seq:35a86fb92842ba47 sfseq:692001 ssnoff:8107faed maplen:41c
> 01:01:10.796 # 
> 01:01:10.801 # netns ns1-600fb815-1a9Kow socket stat for 10003:
> 01:01:10.802 # State Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
> 01:01:10.896 # ESTAB 0      0      [dead:beef:1::1]:37334 [dead:beef:1::1]:10003
> 01:01:10.897 # 	 ts sack cubic wscale:7,7 rto:201 rtt:0.705/0.125 ato:40 mss:65464 pmtu:65536 rcvmss:32768 advmss:65464 cwnd:22 ssthresh:22 bytes_sent:6890524 bytes_acked:6890525 bytes_received:4526163 segs_out:340 segs_in:409 data_segs_out:280 data_segs_in:320 send 16342785816bps lastsnd:30695 lastrcv:30695 lastack:30695 pacing_rate 19607866400bps delivery_rate 787536840bps delivered:281 busy:167ms rcv_rtt:0.459 rcv_space:65476 rcv_ssthresh:65476 minrtt:0.096 tcp-ulp-mptcp flags:Mmec token:0000(id:0)/7eac7c5c(id:0) seq:d4b7487fda3b268d sfseq:450fd4 ssnoff:79893660 maplen:80
> 01:01:10.904 # FAIL: Could not even run loopback v6 test
> 01:01:10.929 not ok 1 selftests: net/mptcp: mptcp_connect.sh # exit=1
> ```
> But not always.

Here the client completely spooled its file and the serve completely acked it. The server instead only send part of its file (4526163  bytes out of 5486620). Anyway the spooled amount of data was completely acked from the other end. Looks like the server missed a wake-up while writing for available write space in poll(). Most likely a bug in the poll() implementation and/or in the subflow wake-up code. (I would wild guess the latter). I'm unsure how to better investigate this beyond code analysis.

**Comment by matttbe** [comment]
@pabeni : here is the output with more details collected by `ss`

```
# INFO: set ns3-601aee64-0QR4Yd dev ns3eth2: ethtool -K tso off gro off
# INFO: set ns4-601aee64-0QR4Yd dev ns4eth3: ethtool -K  gro off
# Created /tmp/tmp.n7kY6CZdEq (size 6839324     /tmp/tmp.n7kY6CZdEq) containing data sent by client
# Created /tmp/tmp.Ynn2TY9Aiq (size 6092828     /tmp/tmp.Ynn2TY9Aiq) containing data sent by server
# New MPTCP socket can be blocked via sysctl            [ OK ]
# setsockopt(..., TCP_ULP, "mptcp", ...) blocked        [ OK ]
# INFO: validating network environment with pings
[   11.290285] netem: version 1.3
# INFO: Using loss of 0.60% delay 42 ms reorder 95% 24% with delay 10ms on ns3eth4
# ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP (duration   146ms) [ OK ]
# ns1 MPTCP -> ns1 (10.0.1.1:10001      ) TCP   (duration   137ms) [ OK ]
# ns1 TCP   -> ns1 (10.0.1.1:10002      ) MPTCP (duration   111ms) [ OK ]
# ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP copyfd_io_poll: poll timed out (events: POLLIN 1, POLLOUT 0)
# (duration 30405ms) [ FAIL ] client exit code 2, server 0
#
# netns ns1-601aee64-0QR4Yd socket stat for 10003:
# Netid State   Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
# tcp   ESTAB   0      0      [dead:beef:1::1]:10003 [dead:beef:1::1]:35166 ino:0 sk:1 <->
#        skmem:(r0,rb6291456,t0,tb2626560,f0,w0,o0,bl0,d0) ts sack cubic wscale:7,7 rto:204 rtt:3.057/6.042 ato:40 mss:32768 pmtu:65536 rcvmss:65436 advmss:65464 cwnd:10 bytes_sent:5699038 bytes_acked:5699038 bytes_received:6839324 segs_out:423 segs_in:367 data_se
gs_out:327 data_segs_in:287 send 857520445bps lastsnd:30324 lastrcv:30576 lastack:30324 pacing_rate 1714970760bps delivery_rate 32768000000bps delivered:328 busy:277ms rwnd_limited:255ms(92.1%) sndbuf_limited:8ms(2.9%) rcv_rtt:0.502 rcv_space:65464 rcv_ssthresh:31
44356 minrtt:0.005 tcp-ulp-mptcp flags:Mec token:0000(id:0)/f6e75411(id:0) seq:44b72c232f2e6c0e sfseq:678001 ssnoff:6a64adb1 maplen:dc1c
# mptcp CLOSING 0      0      [dead:beef:1::1]:10003 [dead:beef:1::1]:35166 timer:(keepalive,29sec,0) ino:0 sk:2 ---
#        skmem:(r0,rb6291456,t0,tb2626560,f1498480,w410256,o0,bl0,d0) remote_key token:f6e75411 write_seq:bd776166f884e1a7 snd_una:bd776166f87ea146 rcv_nxt:44b72c232f2f482b
#
# netns ns1-601aee64-0QR4Yd socket stat for 10003:
# Netid State      Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
# tcp   ESTAB      0      0      [dead:beef:1::1]:35166 [dead:beef:1::1]:10003 ino:0 sk:3 <->
#        skmem:(r0,rb131072,t0,tb2626560,f4096,w0,o0,bl0,d0) ts sack cubic wscale:7,7 rto:201 rtt:0.162/0.096 ato:40 mss:65464 pmtu:65536 rcvmss:32768 advmss:65464 cwnd:10 bytes_sent:6839324 bytes_acked:6839325 bytes_received:5699038 segs_out:367 segs_in:424 data_
segs_out:287 data_segs_in:327 send 32327901235bps lastsnd:30815 lastrcv:30563 lastack:30563 pacing_rate 64506481904bps delivery_rate 1203935632bps delivered:288 busy:29ms rcv_rtt:0.039 rcv_space:65476 rcv_ssthresh:65476 minrtt:0.005 tcp-ulp-mptcp flags:Mmec token:
0000(id:0)/a61f9a60(id:0) seq:bd776166f87e42aa sfseq:569743 ssnoff:9b41f31c maplen:5e9c
# mptcp FIN-WAIT-2 0      0      [dead:beef:1::1]:35166 [dead:beef:1::1]:10003 timer:(keepalive,59sec,0) ino:0 sk:4 ---
#        skmem:(r0,rb131072,t0,tb2626560,f166656,w0,o0,bl0,d0) remote_key token:a61f9a60 write_seq:44b72c232f2f482b snd_una:44b72c232f2f482b rcv_nxt:bd776166f87ea146
# FAIL: Could not even run loopback v6 test
+ rc=1
+ '[' 1 -eq 0 ']'
+ echo 'not ok 1 test: selftests.tap # exit=1'
```

**Comment by matttbe** [comment]
@pabeni : here is the new issue with more debug, I would that helps :)

```
# INFO: set ns3-601d98b0-kMvDTG dev ns3eth2: ethtool -K  gro off
# INFO: set ns4-601d98b0-kMvDTG dev ns4eth3: ethtool -K  gro off
# Created /tmp/tmp.cL2nqwfd5v (size 3424284     /tmp/tmp.cL2nqwfd5v) containing data sent by client
# Created /tmp/tmp.7PH5QXuYPs (size 5301276     /tmp/tmp.7PH5QXuYPs) containing data sent by server
# New MPTCP socket can be blocked via sysctl            [ OK ]
# setsockopt(..., TCP_ULP, "mptcp", ...) blocked        [ OK ]
# INFO: validating network environment with pings
[   11.237438] netem: version 1.3
# INFO: Using loss of 0.62% delay 30 ms reorder 99% 49% with delay 7ms on ns3eth4
# ns1 MPTCP -> ns1 (10.0.1.1:10000      ) MPTCP (duration   146ms) [ OK ]
# ns1 MPTCP -> ns1 (10.0.1.1:10001      ) TCP   (duration   108ms) [ OK ]
# ns1 TCP   -> ns1 (10.0.1.1:10002      ) MPTCP (duration   130ms) [ OK ]
# ns1 MPTCP -> ns1 (dead:beef:1::1:10003) MPTCP (duration   132ms) [ OK ]
# ns1 MPTCP -> ns1 (dead:beef:1::1:10004) TCP   [ FAIL ] file received by client does not match (in, out):
# -rw------- 1 root root 5301276 Feb  5 19:12 /tmp/tmp.7PH5QXuYPs
# Trailing bytes are:
# MPTCP_TEST_FILE_END_MARKER
# -rw------- 1 root root 4759552 Feb  5 19:12 /tmp/tmp.3zpAb8pmmC
# Trailing bytes are:
# E$>{2u!v*CG(duration   114ms) [ FAIL ] client exit code 1, server 0
#
# netns ns1-601d98b0-kMvDTG socket stat for 10004:
# Netid State     Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
# tcp   TIME-WAIT 0      0      [dead:beef:1::1]:10004 [dead:beef:1::1]:37850 timer:(timewait,59sec,0) ino:0 sk:1
#
# TcpActiveOpens                  1                  0.0
# TcpPassiveOpens                 1                  0.0
# TcpInSegs                       499                0.0
# TcpOutSegs                      499                0.0
# TcpExtPruneCalled               18                 0.0
# TcpExtTCPHPHits                 177                0.0
# TcpExtTCPPureAcks               136                0.0
# TcpExtTCPHPAcks                 54                 0.0
# TcpExtTCPBacklogCoalesce        1                  0.0
# TcpExtTCPRcvCoalesce            115                0.0
# TcpExtTCPFromZeroWindowAdv      19                 0.0
# TcpExtTCPToZeroWindowAdv        19                 0.0
# TcpExtTCPWantZeroWindowAdv      55                 0.0
# TcpExtTCPOrigDataSent           321                0.0
# TcpExtTCPDelivered              322                0.0
# MPTcpExtMPCapableFallbackSYNACK 1                  0.0
#
# netns ns1-601d98b0-kMvDTG socket stat for 10004:
# Netid State     Recv-Q Send-Q    Local Address:Port      Peer Address:Port Process
# tcp   TIME-WAIT 0      0      [dead:beef:1::1]:37850 [dead:beef:1::1]:10004 timer:(timewait,59sec,0) ino:0 sk:2
#
# FAIL: Could not even run loopback v6 test
+ rc=1
```

**Comment by matttbe** [comment]
Should be fixed thanks to this patch from @pabeni: 773725f66d5c: mptcp: add a missing retransmission timer scheduling

**Comment by mjmartineau** [comment]
I realize this is closed already - but it also showed up once in kbuild before the patch was merged. Will watch to see if it shows up again.
